### PR TITLE
Multidimensional Resampling

### PR DESCRIPTION
**Pull request by kaifox**
This contains a first (working! ;-) version of multidimensional resampling. Currently there are two available resamplers (a resampler always treats one dimension):

- Repeating
- Linear interpolation (which works only for tensors of field elements)

The simplest starting point is the one in tensorics (which only supports resamplers which do not depend on a field), thus only supports repeating for the moment ;-). Assume a Tensor with two dimensions (String.class, Integer.class). Then one can do now:

`Tensoric<T> resampled = Tensorics.resample(tensor).repeat(String.class).then().repeat(Integer.class).toTensoric();`

If you want to use the interpolation, then one has to use the variant from TensorSupport, e.g.:

`Tensoric<Double> resampled = TensoricDoubles.resample(tensor).repeat(String.class).then().linear(Integer.class, Integer::doubleValue).toTensoric();`

Good starting points to see more examples and to understand the effects are the two tests:

- https://github.com/tensorics/tensorics-core/blob/completion-by-repetition/src/test/org/tensorics/core/tensor/resample/impl/MultiDimensionalResamplingTest.java
- https://github.com/tensorics/tensorics-core/blob/completion-by-repetition/src/test/org/tensorics/core/tensor/resample/impl/LinearInterpolationResamplerTest.java

Please also note the new interface Tensoric<V> which you basically can see as a Tensor, without a fixed shape. Still it has a fixed set of dimensions ;-) (To be considered if more functionality, like dimensionality should go there ...)

In the review/comments, please focus (mainly) on commenting on the general concept, naming, syntax... For little things (e.g. null check here, there ... formatting ... comment)... contributions are very welcome ;-)

Additionally, please look out for loopholes... i.e. it might well be that there should be more checks introduce in the dsl (or another level), ... e.g. it makes no sense to resample a dimension if it is not contained in the tensor... to be seen on which level ...

Otherwise: Any suggestions/comments welcome

### CODE REVIEW

File: src/java/org/tensorics/core/tensor/Positions.java [hunk 1]
```diff
@@ -134,12 +145,16 @@ public Position apply(Position position) {
             return Position.of(builder.build());
         }
 
+        public Position from(Position position) {
```
Comment (michi42) [inline src/java/org/tensorics/core/tensor/Positions.java:148]: I guess this delegation method is just to get a nicer/more fluent syntax elsewhere?

File: src/test/org/tensorics/core/tensor/resample/impl/MultiDimensionalResamplingTest.java [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    }
+
+    @Test
+    public void intThenStringResamplesA10AsB2() {
+        assertThat(intThenString.get("A", 10)).isEqualTo("B2");
+    }
+
+    // @formatter:off
+    /*
+     * This resamples the tensor by repeating first in the Integer and then in the
```
Comment (michi42) [inline src/test/org/tensorics/core/tensor/resample/impl/MultiDimensionalResamplingTest.java]: This one has the comment swapped with the test above...?

**Review by michi42 — APPROVED** [review]
Looks very nice to me! One conceptual question on the Tensoric - I don't think it's clear from the code that a Tensoric is supposed to be bound to a certain dimension set, as the dimension set is by no way obtainable from the interface ... In principle with this interface one could query arbitrary positions (with arbitrary dimensions). Should we formalize this somehow?

File: src/java/org/tensorics/core/tensor/Positions.java (hunk 1 above)
Comment (kaifox) [reply src/java/org/tensorics/core/tensor/Positions.java:148]: Indeed this is the reason ... I thought it is not disturbing too  much ...
you can now write
`Position strippedpos = strip(ACoordinate.class).from(aPosition);`

File: src/test/org/tensorics/core/tensor/resample/impl/MultiDimensionalResamplingTest.java (hunk 2 above)
Comment (kaifox) [reply src/test/org/tensorics/core/tensor/resample/impl/MultiDimensionalResamplingTest.java]: Damn ... Never copy paste ;-) I will correct... Thanks!

**Comment by kaifox** [comment]
I fully agree with Michis comment on the tensoric. I thought in similar ways, but left out this, as it was not immediately necessary ... and was not exactly shure how to implement it. I see the following options:

a) **Extract an interface from shape, which would contain the following methods:**
- `int dimensionality();`
- `contains(...)`
- `Set<Class<?>> dimensions();`  (BTW I would like to introduce this as a replacement for dimensionSet(), as it is more conform to our other namings (and deprecate dimensionSet())
The name of the interface could be e.g. `Dimensional`
(NOTE: If the `contains(..)` method belongs here is not fully clear ... probably better a dedicated interface....

Shape and Tensoric would then implement this... (and thus also the tensor, which might be convenient in any case and simply delegate to shape())

a) **Optionally, we could have a similar interface/object e.g. called `Dimensionality` which could be returned from Tensoric by a dimensionality() method....**

c) **We could strip down the Shape interface (which anyhow should be created) to those methods and e.g.  rename the actual Shape (with the positions() method) to something like a FiniteShape....** 
`Tensoric` would then have a method:
`Shape shape();`
and `Tensor` could then override this by:
`FiniteShape shape();` 

I think I like c) for the moment ... probably with a combination of a) so that we have the convenience methods in Tensor + Tensoric.

... still undesided ... input welcome ;-)

**Pull request merged**